File: python/linked-lists/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        current = head
        count = 1
        groups = [head]
        while current:
            nxt = current.next
            if count % k == 0:
                current.next = None
                groups.append(nxt)
                count = 0
            current = nxt
            count += 1

        def reversal(node):
            if not node:
                return node
            prev, curr, nxt = None, node, node.next

            while nxt:
                curr.next = prev
                prev = curr
                curr = nxt
                nxt = curr.next

            curr.next = prev

            return curr

        finalanswer = ListNode(0)
        final = finalanswer
        for i in groups[:-1]:
            rev = reversal(i)
            final.next = rev
            while final.next:
                final = final.next

        last = groups[-1]
        counter = 0
        while last:
            counter += 1
            last = last.next

        if counter != k:
            final.next = groups[-1]
        else:
            final.next = reversal(groups[-1])
        return finalanswer.next

File: c++/linked-list/reverse-nodes-in-k-group.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        std::vector<ListNode*> groups;
        int count = 0;

        ListNode* group = head;
        ListNode* pointer = head;
        while (pointer) {
            count += 1;
            if (count % k == 0) {
                groups.push_back(group);
                group = pointer->next;
                pointer->next = nullptr;
                pointer = group;
            } else {
                pointer = pointer->next;
            }
        }

        groups.push_back(group);

        ListNode* answer = new ListNode();
        pointer = answer;
        for (ListNode* node : groups) {
            while (pointer->next) {
                pointer = pointer->next;
            }

            if (checkListLength(node) == k) {
                pointer->next = reverseList(node);
            } else {
                pointer->next = node;
            }
        }

        return answer->next;
    }

private:
    int checkListLength(ListNode* head) {
        int length = 0;
        ListNode* pointer = head;
        while (pointer) {
            length += 1;
            pointer = pointer->next;
        }
        return length;
    }

    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }

        ListNode* prev = head;
        ListNode* curr = prev->next;

        while (curr) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        head->next = nullptr;

        return prev;
    }
};
